From 86e6857fb35b2c2e143c354845185d2dc01782ff Mon Sep 17 00:00:00 2001
From: Thibaud Michaud <thibaudm@chromium.org>
Date: Thu, 09 Jan 2025 15:15:18 +0100
Subject: [PATCH] Merged: [wasm][jspi] Add missing entry to wasm builtin list

Add the WasmTrapHandlerThrowTrap builtin to the list of wasm builtins to
ensure that it switches to the central stack before calling the
corresponding runtime function.

R=dlehmann@chromium.org

Bug: 385167053
(cherry picked from commit 91acb7ed025289116087133a04da77102c92861c)

Change-Id: I972b27d1b0f5da1a8c7b680a08d3d2ae5b3f734c
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6163445
Reviewed-by: Daniel Lehmann <dlehmann@chromium.org>
Commit-Queue: Thibaud Michaud <thibaudm@chromium.org>
Cr-Commit-Position: refs/branch-heads/13.2@{#54}
Cr-Branched-From: 24068c59cedad9ee976ddc05431f5f497b1ebd71-refs/heads/13.2.152@{#1}
Cr-Branched-From: 6054ba94db0969220be4f94dc1677fc4696bdc4f-refs/heads/main@{#97085}
---

--- a/v8/src/runtime/runtime-wasm.cc
+++ b/v8/src/runtime/runtime-wasm.cc
@@ -252,6 +252,7 @@
 
 RUNTIME_FUNCTION(Runtime_TrapHandlerThrowWasmError) {
   ClearThreadInWasmScope flag_scope(isolate);
+  CHECK(isolate->IsOnCentralStack());
   HandleScope scope(isolate);
   std::vector<FrameSummary> summary;
   FrameFinder<WasmFrame> frame_finder(isolate, {StackFrame::EXIT});
--- a/v8/src/wasm/wasm-builtin-list.h
+++ b/v8/src/wasm/wasm-builtin-list.h
@@ -146,6 +146,7 @@
   V(DeoptimizationEntry_Eager)                                                 \
   V(WasmLiftoffDeoptFinish)                                                    \
   V(WasmPropagateException)                                                    \
+  V(WasmTrapHandlerThrowTrap)                                                  \
   IF_SHADOW_STACK(V, AdaptShadowStackForDeopt)
 
 namespace detail {
